### ISSUE TITLE
Not returning 404 errors when running production

### ISSUE THREAD

**Issue by pozil** [muenzpraeger/create-lwc-app#106]
### Summary

When serving app in production mode and accessing an invalid URL, there won't be a 404 error and `index.html` will be returned.

### Steps To Reproduce:

1.  Create a new project with sample components
1.  Serve production with `npm run build && npm run serve`
1.  Access an invalid URL: `http://0.0.0.0:3002/invalid`
1.  Notice that `index.html` is returned and there is no 404 error.

### Expected result

There should be a 404 error.

### Actual result

 `index.html` is returned and there is no 404 error.

**Comment by muenzpraeger** [comment]
It‘s an implementation detail up to the individual developer how to handle that in their custom Express config.

**Comment by pozil** [comment]
Ok but default config should be consistent with watch don't you think?

**Comment by muenzpraeger** [comment]
No. Because `watch` is `webpack-dev-server`. `serve` is your own custom Express server.

**Comment by pozil** [comment]
Let's step back for a sec.
This is not the standard express behavior, this is something that is introduced by the tooling.

The problem is here:
https://github.com/muenzpraeger/create-lwc-app/blob/master/packages/lwc-services/src/commands/serve.ts#L93
```ts
app.use('*', (req: any, res: any) => {
    res.sendFile(path.resolve(BUILD_DIR, 'index.html'))
})
```


The `*` implies that all requests that do not match a resource are defaulting to `index.html`.

**Comment by muenzpraeger** [comment]
Yeah, I see. Thought you meant (read too fast) the backend Express option.

What’s wrong with doing a `*`? I think it’s a reasonable fallback.

**Comment by pozil** [comment]
The problem with that default is that you never get a 404 error and that's kind of a blocker for building a REST api.

For example, let's assume that you're building a REST endpoint to retrieve a record in JSON. The endpoint is exposed as `/api/record/:id` where `id` is a record id.

What happens right now if you search for something like `/api/record/invalid_id` you'll get a HTTP 200 response with an an invalid content type (the content of index.html)

**Comment by muenzpraeger** [comment]
Ok, let’s go back to start. In your first example you reference a URI like a webpage. Now you have an API example. These are different things.

Let’s say your scenario is “webpage”. Then IMHO the wildcard fallback is fine.

Let’s say your scenario is “api”. Then the wildcard is only an issue when there is no proper error handling. That would be in your API example to catch within the `/api/record` route the invalid id and return an appropriate error response.

**Comment by pozil** [comment]
I mentioned the webpage example because it was the easiest to reproduce but I've been working on API as well. However you're right about the route handling, I gave a wrong example.

The real problem is when you give a wrong URI whether it is for an API or a webpage.

For example I tried to load an image and got this weird error about invalid content type. That was because my URI was wrong and the server was serving the HTML to the `img` tag instead of the image file.

The other example I had for the REST API was that I was trying to get a resource with an incorrect URI (yes, nothing to do with a bad id, my bad) and I got HTTP 200 with some HTML instead of JSON.

I'm probably bad at writing the right URIs but I don't think that we're helping users by avoiding 404 errors. It just makes debugging harder and again, that's not a common behavior for servers.

**Comment by muenzpraeger** [comment]
So, what would be the right behavior considering that the server serves UI and API at the same time?

I have some ideas in mind to fix that, that may introduce a breaking change. And I’d like first to get your recommendations.

**Comment by pozil** [comment]
I'd say we just go with the default: invalid URI => 404 error.
This is something simple that you see in all websites, users are expecting that.

If you want something nicer, the right way to do this is to throw the 404 error and serve a custom 404 error page with some HTML in the response body when the request expects HTML content type.
Like this: https://trailhead.salesforce.com/invalid

**Comment by muenzpraeger** [comment]
That wouldn’t solve your `api` issue as HTML would be served.

**Comment by pozil** [comment]
Not if you only serve HTML error page when the request has the header: `Accept: text/html`

**Comment by pozil** [comment]
Plus the API caller would know that something is wrong thanks to the 404 error code.

**Comment by muenzpraeger** [comment]
Na, I think it needs a refactor. This discussion showed me that I should redesign the implementation, as it has also other error potential.

**Comment by pozil** [comment]
Ok, thanks for listening. I'm curious about your plan.

FYI I am also struggling with this issue when establishing WebSocket connections (non trivial use case and probably not supported, I know).
What happens is that the server is returning an unexpected HTTP 200 when upgrading the protocol from HTTP to WSS: `app.js:1 WebSocket connection to 'ws://0.0.0.0:3002/' failed: Error during WebSocket handshake: Unexpected response code: 200`
I assume that it's related to that bit of express code since it works on a regular express outside of the LWC project but I need to investigate further to confirm that.

**Comment by pozil** [comment]
Never mind, I figured it out but that has to do with the way express is encapsulated in the LWC tooling that prevents this use case.
For WebSocket.Server to work I need to use the HTTP server that is returned after `express.listen()`. I can't use the `app` parameter of `index.js` module.

I'll just build my own server code, I don't think that you have to worry about that.

**Comment by muenzpraeger** [comment]
That’s exactly the issue. I am looking into spawning a dedicated instance (“client” express on port x, server express on port y), and then proxying.
